feat(HACBS-2293): add a task to update FBC index tag

happybhati opened the pull request:
- This commit adds a new task, `update-ocp-tag` for updating `fromIndex`, `targetIndex`, and `binaryImage`version tags with the version fetched from fbcFragment by `get-ocp-version` task.
- Adds both `get-ocp-version` and `update-ocp-tag` tasks to fbc-release pipeline to accomplish the end results.

theflockers commented on catalog/pipeline/fbc-release/0.18/fbc-release.yaml:
```diff
@@ -0,0 +1,343 @@
+          value: "quay.io/hacbs-release/release-utils:4d8649dbb2b626f5fe9f4ff83c1bc3be268fad31"
+        - name: script
+          value: |
+            set -x
```
I would remove the -x as you are echoing stuff in the end.

theflockers commented on catalog/pipeline/fbc-release/0.18/fbc-release.yaml:
```diff
@@ -0,0 +1,343 @@
+
+            # Function to replace tag in an image
+            replace_tag() {
+                local image="$1"
```
just remember to wrap variables inside `""` with `${}`

happybhati replied:
Thanks for pointing that out. I almost missed that.

mmalina replied:
It's not really needed, in most cases it will work even without curly brackets, but it's generally nicer/safer to use them.

And I think this thread is no longer relevant :)

happybhati replied:
Fixed, thanks @mmalina  and @theflockers

theflockers reviewed (commented):
Also please remember next time to commit the plain copy of the old version to the new one, so we can see the changes easily.

happybhati commented:
```
diff -r catalog/pipeline/fbc-release/0.17 catalog/pipeline/fbc-release/0.18                
diff -r catalog/pipeline/fbc-release/0.17/README.md catalog/pipeline/fbc-release/0.18/README.md
29a30,33
> ### Changes since 0.17
> - added a new task to update `targetIndex`, `fromIndex` and `binaryImage`
>   with provided OCP-Version from the `get-ocp-version` task.
> 
diff -r catalog/pipeline/fbc-release/0.17/fbc-release.yaml catalog/pipeline/fbc-release/0.18/fbc-release.yaml
7c7
<     app.kubernetes.io/version: "0.17"
---
>     app.kubernetes.io/version: "0.18"
132a133,185
>     - name: get-ocp-version
>       taskRef:
>         resolver: "git"
>         params:
>           - name: url
>             value: https://github.com/redhat-appstudio/release-service-bundles.git
>           - name: revision
>             value: main
>           - name: pathInRepo
>             value: catalog/task/get-ocp-version/0.1/get-ocp-version.yaml
>       params:
>         - name: fbcFragment
>           value: $(params.fbcFrgment)
>       workspaces:
>         - name: data
>           workspace: release-workspace
>       runAfter:
>         - verify-enterprise-contract
>     - name: update-version-tag
>       workspaces:
>         - name: input
>           workspace: release-workspace
>       runAfter:
>         - get-ocp-version
>       params:
>         - name: image
>           value: "quay.io/hacbs-release/release-utils:4d8649dbb2b626f5fe9f4ff83c1bc3be268fad31"
>         - name: script
>           value: |
>             fromIndex=$(params.fromIndex)
>             targetIndex=$(params.targetIndex)
>             binaryImage=$(params.binaryImage)
> 
>             # New tag to replace with
>             newTag="$(tasks.get-ocp-version.results.stored-version)"
> 
>             # Function to replace tag in an image
>             replace_tag() {
>                 local image="${1}"
>                 local oldTag=$(echo "$image" | grep -o 'v[0-9]\.*[0-9]{2}$')
>                 local updatedImage="${image/$oldTag/$newTag}"
>                 echo "$updatedImage"
>             }
> 
>             # Access the updated image
>             updatedFromIndex=$(replace_tag "$fromIndex")
>             updatedTargetIndex=$(replace_tag "$targetIndex")
>             updatedBinaryImage=$(replace_tag "$binaryImage")
> 
>             echo "Updated values"
>             echo "$updatedFromIndex" | tee $(results.fromIndex.path)
>             echo "$updatedTargetIndex" | tee $(results.targetIndex.path)
>             echo "$updatedBinaryImage" | tee $(results.binaryImage.path)
156c209
<               value: "$(params.binaryImage)"
---
>               value: "$(tasks.update-version-tag.results.binaryImage)"
158c211
<               value: "$(params.fromIndex)"
---
>               value: "$(tasks.update-version-tag.results.fromIndex)"
261c314
<               value: $(params.targetIndex)
---
>               value: $(tasks.update-version-tag.results.targetIndex)
diff -r catalog/pipeline/fbc-release/0.17/samples/sample_release_PipelineRun.yaml catalog/pipeline/fbc-release/0.18/samples/sample_release_PipelineRun.yaml
54c54
<         value: catalog/pipeline/fbc-release/0.17/fbc-release.yaml
---
>         value: catalog/pipeline/fbc-release/0.18/fbc-release.yaml
diff -r catalog/pipeline/fbc-release/0.17/tests/run.yaml catalog/pipeline/fbc-release/0.18/tests/run.yaml
54c54
<         value: catalog/pipeline/fbc-release/0.17/fbc-release.yaml
---
>         value: catalog/pipeline/fbc-release/0.18/fbc-release.yaml
```

happybhati commented on catalog/pipeline/fbc-release/0.18/fbc-release.yaml:
```diff
@@ -0,0 +1,343 @@
+            # Function to replace tag in an image
+            replace_tag() {
+                local image="$1"
+                local oldTag=$(echo "$image" | grep -o 'v[0-9]*\.[0-9]*')
```
Agree, fixed!

johnbieren commented on catalog/pipeline/fbc-release/0.18/fbc-release.yaml:
```diff
@@ -0,0 +1,341 @@
+            value: catalog/task/get-ocp-version/0.1/get-ocp-version.yaml
+      params:
+        - name: fbcFragment
+          value: $(params.fbcFrgment)
```
did you test this? because this really looks like you didn't considering there is a typo in the variable name

johnbieren replied:
Furthermore, there is no `params.fbcFragment`

happybhati replied:
Thanks for pointing this out, I will fix this, yes I need to test working on fixing my environment.

johnbieren commented on catalog/pipeline/fbc-release/0.18/fbc-release.yaml:
```diff
@@ -0,0 +1,341 @@
+          workspace: release-workspace
+      runAfter:
+        - verify-enterprise-contract
+    - name: update-version-tag
```
this task has no taskref. I have no idea how this is supposed to work

happybhati replied:
Yes, I realized there are better approaches than this. I will add a new task for tag updation and will reference that new task in the release pipeline. Thanks

johnbieren commented on catalog/task/update-ocp-tag/0.1/README.md:
```diff
@@ -0,0 +1,14 @@
+
+Tekton task to update version tag of FBC pull-spec
+ `fromIndex`, `targetIndex` and `binaryImage` with the 
+ provided OCP version from the `get-ocp-version` task.
```
if it is updating it based on a parameter value, the task doesn't really care what other task it came from

johnbieren commented on catalog/task/update-ocp-tag/0.1/README.md:
```diff
@@ -0,0 +1,14 @@
+|------|-------------|----------|---------------|
+| fromIndex | The source Index image (catalog of catalogs) FBC fragment | No | - |
+| targetIndex | Index image (catalog of catalogs) the FBC fragment will be added to | No | - |
+| binaryImage | OCP binary image to be baked into the index image | No | "" |
```
if it isn't optional why is there a Default value?

johnbieren commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,66 @@
+       #!/usr/bin/env sh
+       set -eux
+
+        fromIndex="$(params.fromIndex)"
```
I am confused why we do this instead of just use the param values directly on lines 59-61

johnbieren commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,66 @@
+        binaryImage="$(params.binaryImage)"
+
+        # New tag to replace with
+        newTag="$(params.ocpVersion)"
```
why not just use the parameter directly and not add this line

johnbieren commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,66 @@
+      image: quay.io/hacbs-release/release-utils:4d8649dbb2b626f5fe9f4ff83c1bc3be268fad31
+      script: |
+       #!/usr/bin/env sh
+       set -eux
```
we have `echo` and `tee` do we need `set -x`?

johnbieren replied:
You may have still wanted the `set -eu` not sure

mmalina replied:
I agree with Johnny, `set -eu` is probably a good idea.

johnbieren commented on catalog/pipeline/fbc-release/0.18/README.md:
```diff
@@ -0,0 +1,115 @@
+
+### Changes since 0.17
+- adds new tasks `get-ocp-version` and `update-ocp-tag` to update version tag
+  for `targetIndex`, `fromIndex` and `binaryImage` with valid OCP-Version.
```
no `.` at the end to match the other lines

johnbieren commented on catalog/pipeline/fbc-release/0.18/fbc-release.yaml:
```diff
@@ -0,0 +1,330 @@
+      workspaces:
+        - name: data
+          workspace: release-workspace
+      runAfter:
```
you don't need this. You pass a result of `get-ocp-version` as a parameter to this task, so tekton knows it needs to run after it

happybhati commented:
```
diff -r catalog/pipeline/fbc-release/0.17/README.md catalog/pipeline/fbc-release/0.18/README.md
29a30,33
> ### Changes since 0.17
> - adds new tasks `get-ocp-version` and `update-ocp-tag` to update version tag
>   for `targetIndex`, `fromIndex` and `binaryImage` with valid OCP-Version
> 
diff -r catalog/pipeline/fbc-release/0.17/fbc-release.yaml catalog/pipeline/fbc-release/0.18/fbc-release.yaml
7c7
<     app.kubernetes.io/version: "0.17"
---
>     app.kubernetes.io/version: "0.18"
132a133,172
>     - name: get-ocp-version
>       taskRef:
>         resolver: "git"
>         params:
>           - name: url
>             value: https://github.com/redhat-appstudio/release-service-bundles.git
>           - name: revision
>             value: main
>           - name: pathInRepo
>             value: catalog/task/get-ocp-version/0.1/get-ocp-version.yaml
>       params:
>         - name: fbcFragment
>           value: "$(tasks.collect-data.results.fbcFragment)"
>       workspaces:
>         - name: data
>           workspace: release-workspace
>       runAfter:
>         - verify-enterprise-contract
>     - name: update-ocp-tag
>       taskRef:
>         resolver: "git"
>         params:
>           - name: url
>             value: https://github.com/redhat-appstudio/release-service-bundles.git
>           - name: revision
>             value: main
>           - name: pathInRepo
>             value: catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml
>       params:
>         - name: fromIndex
>           value: "$(params.fromIndex)"
>         - name: targetIndex
>           value: "$(params.targetIndex)"
>         - name: binaryImage
>           value: "$(params.binaryImage)"
>         - name: ocpVersion
>           value: "$(tasks.get-ocp-version.results.stored-version)"
>       workspaces:
>         - name: data
>           workspace: release-workspace
156c196
<               value: "$(params.binaryImage)"
---
>               value: "$(tasks.update-ocp-tag.results.updated-binaryImage)"
158c198
<               value: "$(params.fromIndex)"
---
>               value: "$(tasks.update-ocp-tag.results.updated-fromIndex)"
261c301
<               value: $(params.targetIndex)
---
>               value: $(tasks.update-ocp-tag.results.updated-targetIndex)
diff -r catalog/pipeline/fbc-release/0.17/samples/sample_release_PipelineRun.yaml catalog/pipeline/fbc-release/0.18/samples/sample_release_PipelineRun.yaml
54c54
<         value: catalog/pipeline/fbc-release/0.17/fbc-release.yaml
---
>         value: catalog/pipeline/fbc-release/0.18/fbc-release.yaml
diff -r catalog/pipeline/fbc-release/0.17/tests/run.yaml catalog/pipeline/fbc-release/0.18/tests/run.yaml
54c54
<         value: catalog/pipeline/fbc-release/0.17/fbc-release.yaml
---
>         value: catalog/pipeline/fbc-release/0.18/fbc-release.yaml
```

kasemAlem commented on catalog/pipeline/fbc-release/0.18/README.md:
```diff
@@ -0,0 +1,115 @@
+
+## Changelog
+
+### Changes since 0.17
```
the current version is [0.18](https://github.com/redhat-appstudio/release-service-bundles/blob/main/catalog/pipeline/fbc-release/0.18/fbc-release.yaml#L7C17-L7C17)    , I think you need to make it 0.19 .

kasemAlem commented on catalog/pipeline/fbc-release/0.18/fbc-release.yaml:
```diff
@@ -0,0 +1,328 @@
+metadata:
+  name: fbc-release
+  labels:
+    app.kubernetes.io/version: "0.18"
```
```suggestion
    app.kubernetes.io/version: "0.19"
```

theflockers commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,66 @@
+
+        # New tag to replace with
+        # newTag="$(params.ocpVersion)"
+
```
remove the commented lines.

johnbieren commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,58 @@
+      type: string
+    - name: binaryImage
+      description: OCP binary image to be baked into the index image
+      default: ""
```
shouldn't have default value

johnbieren commented on catalog/pipeline/fbc-release/0.19/README.md:
```diff
@@ -0,0 +1,121 @@
+
+### Changes since 0.18
+- adds new tasks `get-ocp-version` and `update-ocp-tag` to update version tag
+  for `targetIndex`, `fromIndex` and `binaryImage` with valid OCP-Version
```
no `-` .. OCP version

mmalina commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,57 @@
+      type: string
+  results:
+    - name: updated-fromIndex
+      description: Store OCP version number from given Image
```
What does this mean? Does it store the ocp version? In that case just `OCP version number from given Image` would be more suitable. This applies to the others as well.

mmalina replied:
But looking at what it does, it's exactly the same as the source params, only the tag is modified to include the new ocpVersion, right? So the description should say that.

mmalina commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,57 @@
+        replace_tag() {
+            local image="${1}"
+            local oldTag=$(echo "$image" | sed -E 's/.*([vV][0-9]+\.[0-9]{2})$/\1/')
+            local updatedImage="${image/$oldTag/"$(params.ocpVersion)"}"
```
I'd suggest a much simpler way - using bash string substitution:
```
replace_tag() {
    echo "${1%:*}:$(params.ocpVersion)"
}
```

theflockers replied:
This might be better indeed. Noticed that Scott's images does not have a tag in the given format, but `:latest`.

mmalina replied:
The two lines above are no longer used and should be removed.

happybhati replied:
Done, thanks for pointing out @mmalina

mmalina replied:
Cool

mmalina commented on catalog/pipeline/fbc-release/0.19/fbc-release.yaml:
```diff
@@ -0,0 +1,344 @@
+          value: "$(params.binaryImage)"
+        - name: ocpVersion
+          value: "$(tasks.get-ocp-version.results.stored-version)"
+      workspaces:
```
AFAICS, the task doesn't even define a workspace, so I think this should be removed.

theflockers replied:
Yes. I tried the pipeline and it fails because of this.

theflockers commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,57 @@
+        echo "Updated values"
+        echo "$updatedFromIndex" | tee $(results.updated-fromIndex.path)
+        echo "$updatedTargetIndex" | tee $(results.updated-targetIndex.path)
+        echo "$updatedBinaryImage" | tee $(results.updated-binaryImage.path)
```
use `echo -n` instead as it is adding "\n"  to the result and breaking the pipelinerun.
```json
- name: params
      value: >-
        [{"name":"binaryImage","value":"registry.redhat.io/openshift4/ose-operator-registry:v4.12

        "},{"name":"fromIndex","value":"quay.io/scoheb/fbc-index-testing:v4.12

        "},
```

happybhati replied:
Done, thanks for pointing out @theflockers

theflockers replied:
So, it works fine now but the output is all in the same line (as `tee` only prints out what it receives).
In the sense of functionality it is all fine, but I guess that we want a readable output. 

Maybe change it to a regular `echo -n XXX > $(path.to.result)`  and go back adding the `set -x`? @mmalina @johnbieren

mmalina replied:
Yeah, that, or you could also just add a plain `echo` after each line.

happybhati replied:
Thanks for the input @theflockers and @mmalina 
updated.

mmalina commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,57 @@
+      description: OCP binary image to be baked into the index image
+      type: string
+    - name: ocpVersion
+      description: OCP version tag to replace the current set tags on indexe images
```
Typo: indexe -> index

happybhati replied:
Fixed, thanks :)

theflockers commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,59 @@
+
+        echo "Updated values"
+        echo -n "$updatedFromIndex" | tee $(results.updated-fromIndex.path)
+        echo    # Print a new line
```
I think these comments are redundant. You can remove them.

mmalina replied:
@theflockers they were added because of this thread you created :) https://github.com/redhat-appstudio/release-service-bundles/pull/175#discussion_r1311813663

happybhati replied:
+1 @mmalina

theflockers replied:
The echo is fine. Just saying that it is pretty clear what this `echo` it does and that the comments are unnecessary, but this is opinion. In Germany with my opinions + 2.80 I can ride around by bus.

mmalina replied:
Ah, I read "comments" as "commands" :)

johnbieren commented:
fbc-release 0.19 now exists so this will probably need to be rebased to create a 0.20

happybhati committed:
feat(HACBS-2293): add a task to update FBC index tag

- This commit adds a new task, `update-ocp-tag` for updating
 `fromIndex`, `targetIndex`, and `binaryImage`version tags with
 the version fetched from fbcFragment by `get-ocp-version` task.

- Adds both `get-ocp-version` and `update-ocp-tag` tasks to
  fbc-release pipeline to accomplish the end results.

Signed-off-by: Happy Bhati <hbhati@redhat.com>

theflockers commented:
LGTM  but needs rebase and new version as @johnbieren commented.

happybhati commented:
> LGTM  but needs rebase and new version as @johnbieren commented.

@theflockers I already did rebase and added new version.
@mmalina and @johnbieren if you can approve please, thanks.

mmalina commented on catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml:
```diff
@@ -0,0 +1,57 @@
+        echo    # Print a new line
+        echo -n "$updatedTargetIndex" | tee $(results.updated-targetIndex.path)
+        echo    # Print a new line
+        echo -n "$updatedBinaryImage" | tee $(results.updated-binaryImage.path)
```
I'd suggest printing another line after this as well - it's always good to end any output with "\n".

mmalina commented:
> > LGTM  but needs rebase and new version as @johnbieren commented.
> 
> @theflockers I already did rebase and added new version. @mmalina and @johnbieren if you can approve please, thanks.

Except for the small suggestion [to print another new line at the end](https://github.com/redhat-appstudio/release-service-bundles/pull/175#discussion_r1315458006), I'm fine with the changes.

johnbieren reviewed (changes requested):
0.19 has changed since you created 0.20 from it. Can you redo it? the ec bundle stuff is gone

johnbieren commented on catalog/pipeline/fbc-release/0.20/README.md:
```diff
@@ -0,0 +1,124 @@
+| fbcPublishingCredentials | Secret used to publish the built index image | Yes | "fbc-publishing-credentials" |
+| requestUpdateTimeout | Max seconds to wait until the status is updated | Yes | - |
+| buildTimeoutSeconds | Max seconds to wait until the build finishes | Yes | - |
+| verify_ec_task_bundle | The location of the bundle containing the verify-enterprise-contract task | No | - |
```
this is not present in 0.19

happybhati replied:
Thanks for pointing it out @johnbieren 
Now fixed!

happybhati commented:
```
diff -r catalog/pipeline/fbc-release/0.19/README.md catalog/pipeline/fbc-release/0.20/README.md
31a32,35
> ### Changes since 0.19
> - adds new tasks `get-ocp-version` and `update-ocp-tag` to update version tag
>   for `targetIndex`, `fromIndex` and `binaryImage` with valid OCP Version
> 
diff -r catalog/pipeline/fbc-release/0.19/fbc-release.yaml catalog/pipeline/fbc-release/0.20/fbc-release.yaml
7c7
<     app.kubernetes.io/version: "0.19"
---
>     app.kubernetes.io/version: "0.20"
146a147,180
>     - name: get-ocp-version
>       taskRef:
>         resolver: "git"
>         params:
>           - name: url
>             value: https://github.com/redhat-appstudio/release-service-bundles.git
>           - name: revision
>             value: main
>           - name: pathInRepo
>             value: catalog/task/get-ocp-version/0.1/get-ocp-version.yaml
>       params:
>         - name: fbcFragment
>           value: "$(tasks.collect-data.results.fbcFragment)"
>       runAfter:
>         - verify-enterprise-contract
>     - name: update-ocp-tag
>       taskRef:
>         resolver: "git"
>         params:
>           - name: url
>             value: https://github.com/redhat-appstudio/release-service-bundles.git
>           - name: revision
>             value: main
>           - name: pathInRepo
>             value: catalog/task/update-ocp-tag/0.1/update-ocp-tag.yaml
>       params:
>         - name: fromIndex
>           value: "$(params.fromIndex)"
>         - name: targetIndex
>           value: "$(params.targetIndex)"
>         - name: binaryImage
>           value: "$(params.binaryImage)"
>         - name: ocpVersion
>           value: "$(tasks.get-ocp-version.results.stored-version)"
172c206
<               value: "$(params.binaryImage)"
---
>               value: "$(tasks.update-ocp-tag.results.updated-binaryImage)"
174c208
<               value: "$(params.fromIndex)"
---
>               value: "$(tasks.update-ocp-tag.results.updated-fromIndex)"
281c315
<               value: $(params.targetIndex)
---
>               value: $(tasks.update-ocp-tag.results.updated-targetIndex)
diff -r catalog/pipeline/fbc-release/0.19/samples/sample_release_PipelineRun.yaml catalog/pipeline/fbc-release/0.20/samples/sample_release_PipelineRun.yaml
58c58
<         value: catalog/pipeline/fbc-release/0.19/fbc-release.yaml
---
>         value: catalog/pipeline/fbc-release/0.20/fbc-release.yaml
diff -r catalog/pipeline/fbc-release/0.19/tests/run.yaml catalog/pipeline/fbc-release/0.20/tests/run.yaml
58c58
<         value: catalog/pipeline/fbc-release/0.19/fbc-release.yaml
---
>         value: catalog/pipeline/fbc-release/0.20/fbc-release.yaml

```
